File: src/main/java/com/anod/appwatcher/ui/DrawerActivity.java
package com.anod.appwatcher.ui;

import android.accounts.Account;
import android.content.Intent;
import android.database.ContentObserver;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.text.format.DateUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.anod.appwatcher.MarketSearchActivity;
import com.anod.appwatcher.Preferences;
import com.anod.appwatcher.R;
import com.anod.appwatcher.SettingsActivity;
import com.anod.appwatcher.accounts.AccountChooser;
import com.anod.appwatcher.content.DbContentProvider;
import com.anod.appwatcher.content.TagsContentProviderClient;
import com.anod.appwatcher.content.TagsCursor;
import com.anod.appwatcher.fragments.AccountChooserFragment;
import com.anod.appwatcher.installed.ImportInstalledActivity;
import com.anod.appwatcher.model.Tag;
import com.anod.appwatcher.tags.AppsTagActivity;
import com.anod.appwatcher.wishlist.WishlistFragment;

/**
 * @author alex
 * @date 2014-08-07
 */
abstract public class DrawerActivity extends ToolbarActivity implements AccountChooser.OnAccountSelectionListener {
    DrawerLayout mDrawerLayout;
    private TextView mAccountNameView;
    protected NavigationView mNavigationView;
    private AccountChooser mAccountChooser;
    protected Preferences mPreferences;
    private String mAuthToken;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPreferences = new Preferences(this);
        mAccountChooser = new AccountChooser(this, mPreferences, this);
        mAccountChooser.init();
    }

    protected void setupDrawer() {
        setupToolbar();

        if (isDrawerEnabled()) {
            mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
            mNavigationView = (NavigationView) findViewById(R.id.nav_view);
            setupDrawerContent(mNavigationView);
            updateTags();
            getContentResolver().registerContentObserver(DbContentProvider.TAGS_CONTENT_URI, true, new TagsUpdateObserver(this));
        }
    }

    private void setupDrawerContent(NavigationView navigationView) {
        View headerView = navigationView.getHeaderView(0);

        mAccountNameView = ((TextView) headerView.findViewById(R.id.account_name));
        LinearLayout changeAccount = (LinearLayout) headerView.findViewById(R.id.account_change);
        changeAccount.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onAccountChooseClick();
            }
        });

        Preferences pref = new Preferences(this);
        long time = pref.getLastUpdateTime();

        TextView lastUpdateView = (TextView) headerView.findViewById(R.id.last_update);
        if (time > 0) {
            String lastUpdate = getString(R.string.last_update, DateUtils.getRelativeDateTimeString(this, time, DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0));
            lastUpdateView.setText(lastUpdate);
            lastUpdateView.setVisibility(View.VISIBLE);
        } else {
            lastUpdateView.setVisibility(View.GONE);/**/
        }

        navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(MenuItem menuItem) {
                        mDrawerLayout.closeDrawers();
                        onOptionsItemSelected(menuItem);
                        return true;
                    }
                });

    }

    public void updateTags() {
        Menu menu = mNavigationView.getMenu();
        menu.removeGroup(1);

        TagsContentProviderClient tags = new TagsContentProviderClient(this);
        TagsCursor cr = tags.queryAll();
        if (cr == null || !cr.moveToFirst()) {
            return;
        }
        cr.moveToPosition(-1);

        while (cr.moveToNext()) {
            Tag tag = cr.getTag();
            MenuItem item = menu.add(1, tag.id, tag.id, tag.name);
            item.setActionView(R.layout.drawer_tag_indicator);
            ImageView iv = (ImageView) item.getActionView();
            Drawable d = ResourcesCompat.getDrawable(getResources(), R.drawable.circular_color, null);
            DrawableCompat.setTint(d, tag.color);
            iv.setImageDrawable(d);
            item.setIntent(AppsTagActivity.createTagIntent(tag, this));
        }
        cr.close();
    }

    @Override
    public AccountChooserFragment.OnAccountSelectionListener getAccountSelectionListener() {
        return mAccountChooser;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        mAccountChooser.onRequestPermissionResult(requestCode, permissions, grantResults);
    }

    protected void onAccountChooseClick() {
        mAccountChooser.showAccountsDialogWithCheck();
    }

    @Override
    public void onAccountSelected(Account account, String authToken) {
        mAuthToken = authToken;
        if (authToken == null) {
            Toast.makeText(this, R.string.failed_gain_access, Toast.LENGTH_LONG).show();
            return;
        }

        if (isDrawerEnabled()) {
            setDrawerAccount(account);
        }
    }

    @Override
    public void onAccountNotFound() {
        Toast.makeText(this, R.string.failed_gain_access, Toast.LENGTH_LONG).show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isDrawerEnabled()) {
                    mDrawerLayout.openDrawer(GravityCompat.START);
                    return true;
                }
                return super.onOptionsItemSelected(item);
            case R.id.menu_add:
                Intent addActivity = new Intent(this, MarketSearchActivity.class);
                startActivity(addActivity);
                return true;
            case R.id.menu_settings:
                Intent settingsActivity = new Intent(this, SettingsActivity.class);
                startActivity(settingsActivity);
                return true;
            case R.id.menu_act_import:
                startActivity(new Intent(this, ImportInstalledActivity.class));
                return true;
            case R.id.menu_wishlist:
                Bundle args = new Bundle();
                args.putParcelable(WishlistFragment.EXTRA_ACCOUNT, mAccountChooser.getAccount());
                args.putString(WishlistFragment.EXTRA_AUTH_TOKEN, mAuthToken);
                startActivity(FragmentToolbarActivity.intent(WishlistFragment.TAG, args, this));
                return true;
        }
        Intent intent = item.getIntent();
        if (intent != null) {
            startActivity(intent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    protected boolean isDrawerEnabled() {
        return true;
    }

    protected void setDrawerAccount(Account account) {
        if (account == null) {
            mAccountNameView.setText(R.string.choose_an_account);
        } else {
            mAccountNameView.setText(account.name);
        }
    }

    public boolean isAuthenticated() {
        return mAuthToken != null;
    }

    public void showAccountsDialogWithCheck() {
        Toast.makeText(this, R.string.failed_gain_access, Toast.LENGTH_LONG).show();
        mAccountChooser.showAccountsDialogWithCheck();
    }

    static class TagsUpdateObserver extends ContentObserver {
        private final DrawerActivity mDrawerActivity;

        TagsUpdateObserver(DrawerActivity drawerActivity) {
            super(new Handler());
            mDrawerActivity = drawerActivity;
        }

        @Override
        public void onChange(boolean selfChange) {
            mDrawerActivity.updateTags();
        }
    }
}

File: src/main/java/com/anod/appwatcher/ui/DrawerActivity.kt
package com.anod.appwatcher.ui

import android.accounts.Account
import android.content.Intent
import android.database.ContentObserver
import android.os.Bundle
import android.os.Handler
import android.support.design.widget.NavigationView
import android.support.v4.content.res.ResourcesCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.text.format.DateUtils
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import com.anod.appwatcher.*
import com.anod.appwatcher.accounts.AccountSelectionDialog
import com.anod.appwatcher.accounts.AuthTokenAsync
import com.anod.appwatcher.accounts.AuthTokenBlocking
import com.anod.appwatcher.content.DbContentProvider
import com.anod.appwatcher.content.TagsContentProviderClient
import com.anod.appwatcher.installed.ImportInstalledActivity
import com.anod.appwatcher.tags.AppsTagActivity
import com.anod.appwatcher.wishlist.WishlistFragment

/**
 * @author alex
 * *
 * @date 2014-08-07
 */
abstract class DrawerActivity : ToolbarActivity(), AccountSelectionDialog.SelectionListener {

    private var drawerLayout: DrawerLayout? = null
    private var accountNameView: TextView? = null
    private var navigationView: NavigationView? = null
    private var authToken: String? = null

    private val accountSelectionDialog: AccountSelectionDialog by lazy {
        AccountSelectionDialog(this, App.provide(this).prefs, this)
    }

    open val isHomeAsMenu: Boolean
        get() = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val account = App.provide(this).prefs.account
        if (account== null) {
            accountSelectionDialog.show()
        } else {
            onAccountSelected(account)
        }
    }

    protected fun setupDrawer() {
        setupToolbar()

        this.drawerLayout = findViewById(R.id.drawer_layout)
        this.navigationView = findViewById(R.id.nav_view)

        if (this.navigationView != null) {
            this.drawerLayout?.addDrawerListener(object: DrawerLayout.SimpleDrawerListener() {
                override fun onDrawerOpened(drawerView: View?) {
                    super.onDrawerOpened(drawerView)
                    drawerView?.postDelayed({
                        updateTags()
                    }, 300L)
                }
            })

            setupDrawerContent()
            updateTags()
            val observer = TagsUpdateObserver(this)
            contentResolver.registerContentObserver(DbContentProvider.tagsUri, true, observer)
            contentResolver.registerContentObserver(DbContentProvider.appsTagUri, true, observer)
        }
    }

    private fun setupDrawerContent() {
        val headerView = this.navigationView?.getHeaderView(0) ?: return

        accountNameView = headerView.findViewById<View>(R.id.account_name) as TextView
        val changeAccount = headerView.findViewById<View>(R.id.account_change) as LinearLayout
        changeAccount.setOnClickListener {
            this.accountSelectionDialog.show()
        }

        val time = App.provide(this).prefs.lastUpdateTime

        val lastUpdateView = headerView.findViewById<View>(R.id.last_update) as TextView
        if (time > 0) {
            val lastUpdate = getString(R.string.last_update, DateUtils.getRelativeDateTimeString(this, time, DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0))
            lastUpdateView.text = lastUpdate
            lastUpdateView.visibility = View.VISIBLE
        } else {
            lastUpdateView.visibility = View.GONE
        }

        this.navigationView?.setNavigationItemSelectedListener { menuItem ->
            this.drawerLayout?.closeDrawers()
            onOptionsItemSelected(menuItem)
            true
        }
    }

    fun updateTagsIfVisible() {
        if (this.drawerLayout?.isDrawerVisible(GravityCompat.START) == true) {
            updateTags()
        }
    }

    private fun updateTags() {
        val menu = this.navigationView?.menu ?: return
        menu.removeGroup(1)

        val tags = TagsContentProviderClient(this)
        val counts = tags.queryTagsAppsCounts()

        val cr = tags.queryAll()
        cr.moveToPosition(-1)
        while (cr.moveToNext()) {
            val tag = cr.tag
            val count = counts.get(tag.id)
            val item = menu.add(1, tag.id, tag.id, tag.name)
            item.setActionView(R.layout.drawer_tag_indicator)
            val tagIndicator = item.actionView.findViewById<View>(android.R.id.text1) as TextView
            val d = ResourcesCompat.getDrawable(resources, R.drawable.circular_color, null)
            DrawableCompat.setTint(DrawableCompat.wrap(d!!), tag.color)
            tagIndicator.setBackgroundDrawable(d)
            tagIndicator.text = if (count > 99) "99+" else "" + count
            item.intent = AppsTagActivity.createTagIntent(tag, this)
        }
        cr.close()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        accountSelectionDialog.onRequestPermissionResult(requestCode, permissions, grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        accountSelectionDialog.onActivityResult(requestCode, resultCode, data)
    }

    override fun onAccountSelected(account: Account) {
        AuthTokenAsync(this).request(this, account, object : AuthTokenAsync.Callback {
            override fun onToken(token: String) {
                authToken = token
                setDrawerAccount(account)
            }

            override fun onError(errorMessage: String) {
                if (App.with(this@DrawerActivity).isNetworkAvailable) {
                    Toast.makeText(this@DrawerActivity, R.string.failed_gain_access, Toast.LENGTH_LONG).show()
                } else {
                    Toast.makeText(this@DrawerActivity, R.string.check_connection, Toast.LENGTH_SHORT).show()
                }
                return
            }
        })
    }

    override fun onAccountNotFound(errorMessage: String) {
        if (App.with(this).isNetworkAvailable) {
            if (errorMessage.isNotBlank()) {
                Toast.makeText(this, errorMessage, Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(this, R.string.failed_gain_access, Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(this, R.string.check_connection, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                if (this.isHomeAsMenu) {
                    this.drawerLayout?.openDrawer(GravityCompat.START)
                    return true
                }
                return super.onOptionsItemSelected(item)
            }
            R.id.menu_add -> {
                val addActivity = Intent(this, MarketSearchActivity::class.java)
                startActivity(addActivity)
                return true
            }
            R.id.menu_settings -> {
                val settingsActivity = Intent(this, SettingsActivity::class.java)
                startActivity(settingsActivity)
                return true
            }
            R.id.menu_act_import -> {
                startActivity(Intent(this, ImportInstalledActivity::class.java))
                return true
            }
            R.id.menu_wishlist -> {
                val args = Bundle()
                args.putParcelable(WishlistFragment.EXTRA_ACCOUNT, App.provide(this).prefs.account)
                args.putString(WishlistFragment.EXTRA_AUTH_TOKEN, authToken)
                startActivity(FragmentToolbarActivity.intent(WishlistFragment.TAG, args, this))
                return true
            }
        }
        val intent = item.intent
        if (intent != null) {
            startActivity(intent)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setDrawerAccount(account: Account?) {
        if (account == null) {
            accountNameView?.setText(R.string.choose_an_account)
        } else {
            accountNameView?.text = account.name
        }
    }

    val isAuthenticated: Boolean
        get() = authToken != null

    fun showAccountsDialogWithCheck() {
        Toast.makeText(this, R.string.failed_gain_access, Toast.LENGTH_LONG).show()
        accountSelectionDialog.show()
    }

    internal class TagsUpdateObserver(private val activity: DrawerActivity) : ContentObserver(Handler()) {
        override fun onChange(selfChange: Boolean) {
            activity.updateTagsIfVisible()
        }
    }
}
